docs/reference/query-languages/esql/esql-commands.md

```diff
@@ -663,6 +663,72 @@ FROM employees
 | LIMIT 5
 ```
 
+## `LOOKUP JOIN` [esql-lookup-join]
+
+::::{warning}
+This functionality is in technical preview and may be changed or removed in a future release. Elastic will work to fix any issues, but features in technical preview are not subject to the support SLA of official GA features.
+::::
+
+`LOOKUP JOIN` enables you to add data from another index, AKA a 'lookup' index, to your {{esql}} query results, simplifying data enrichment and analysis workflows.
+
+**Syntax**
+
+```
+FROM <source_index>
+| LOOKUP JOIN <lookup_index> ON <field_name>
+```
+
+```esql
+FROM firewall_logs
+| LOOKUP JOIN threat_list ON source.IP
+| WHERE threat_level IS NOT NULL
+```
+
+**Parameters**
+
+`<lookup_index>`
+: The name of the lookup index. This must be a specific index name - wildcards, aliases, and remote cluster references are not supported.
+
+`<field_name>`
+: The field to join on. This field must exist in both your current query results and in the lookup index. If the field contains multi-valued entries, those entries will not match anything (the added fields will contain `null` for those rows).
+
+**Description**
+
+The `LOOKUP JOIN` command adds new columns to your {esql} query results table by finding documents in a lookup index that share the same join field value as your result rows.
+
+For each row in your results table that matches a document in the lookup index based on the join field, all fields from the matching document are added as new columns to that row.
+
+If multiple documents in the lookup index match a single row in your results, the output will contain one row for each matching combination.
+
+**Examples**
+
+**IP Threat correlation**: This query would allow you to see if any source IPs match known malicious addresses.
+
+```esql
+FROM firewall_logs
+| LOOKUP JOIN threat_list ON source.IP
+```
+
+**Host metadata correlation**: This query pulls in environment or ownership details for each host to correlate with your metrics data.
+
+```esql
+FROM system_metrics
+| LOOKUP JOIN host_inventory ON host.name
+| LOOKUP JOIN employees ON host.name
+```
+
+**Service ownership mapping**: This query would show logs with the owning team or escalation information for faster triage and incident response.
+
+```esql
+FROM app_logs
+| LOOKUP JOIN service_owners ON service_id
+```
+
+In case of name collisions, the newly created columns will override existing columns.
+
+```esql
+
```

Comment:
```suggestion
FROM Left
| WHERE Language IS NOT NULL // works and filter TLD UK
| LOOKUP JOIN Right ON Key
// same semantics and result when moving the WHERE clause after the filter
// in fact the optimizer will move the filter before the lookup
alex-spies marked this conversation as resolved.
FROM Left
| LOOKUP JOIN Right ON Key
| WHERE Language IS NOT NULL 
```

Reply:
This is useful to mention, maybe with
"`LOOKUP JOIN`s are generally faster when there are fewer rows to join with. ES|QL will try and perform any `WHERE` commands before the `LOOKUP JOIN` where possible; for instance, in the following example it is not needed to move the `WHERE` command manually ahead of the `LOOKUP JOIN`:"

OTOH, we also have an [open issue](https://github.com/elastic/elasticsearch/issues/117766) to give the optimizer an own page in the docs, where this could also be mentioned. It doesn't hurt to call out here, though.